### DISCUSSION TITLE
With WCSPH, why is the density limited *after* using it in the viscosity?

### DISCUSSION THREAD

**Discussion by efaulhaber** [InteractiveComputerGraphics/SPlisHSPlasH#277] [Q&A]
With WCSPH, the density is limited to not be lower than the rest density of the fluid, resulting in a non-negative pressure.
This limiting happens here:
https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/8454e9f454fef20771dfe98d318904da62764b4c/SPlisHSPlasH/WCSPH/TimeStepWCSPH.cpp#L105
However, the non-limited density is already used earlier to calculate the viscosity, here:
https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/8454e9f454fef20771dfe98d318904da62764b4c/SPlisHSPlasH/WCSPH/TimeStepWCSPH.cpp#L92

Is there a specific reason for not using the limited density in the viscosity formulation?

**Comment by janbender** [comment]
In SPH the density is often used to determine the volume (mass / density) that is represented by a particle. This volume is used in the SPH sum (e.g. when computing viscosity). However, when computing the pressure values the density value is used in another way. And the density value at the free surface leads to a negative pressure due to particle deficiency in the neighborhood of a surface particle. This can cause a particle clumping at the surface. To avoid this a common solution is the so-called pressure clamping to positive pressure values. In our implementation we clamped the density since it is not used anymore by other components. However, if you prefer, you could also clamp the pressure instead after line 106 which yields the same result. 

Hope this helps to understand the implementation.

**Reply by efaulhaber** [reply]
Thanks, that helps! So using the clamped density for other computations like viscosity would be incorrect, then?

**Reply by efaulhaber** [reply]
Also, with clamped pressure, there is also particle clumping happening at the surface. It is just happening slowly and not as rapidly as with attractive forces between surface particles. Or am I misunderstanding things here?

**Comment by janbender** [comment]
This should not happen since there is no attraction force between the particles if you clamp the pressure. Do you use surface tension? Then you have an attraction force.

**Reply by efaulhaber** [reply]
But there is also no repulsive force between particles at the free surface. Without the attractive forces, they just need longer to settle in clumped positions:
![splishsplash_tank](https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/assets/44124897/5ffd4340-c6a2-4688-8d72-6610387cc118)

**Comment by janbender** [comment]
The problem at the surface is that due to particle deficiency in the neighborhood of a surface particle the repulsive forces are not strong enough the push the particles apart. I would assume a larger stiffness coefficient could help a bit. However, anyway I would recommend to use an implicit solver like DFSPH since this avoid stability issues and the vibrations in your simulation.

**Reply by efaulhaber** [reply]
Thanks a lot! We just started working with SPH and writing our own code, so we naturally started studying WCSPH first. We can now reproduce SPlisHSPlasH WCSPH simulations with our code, and implicit solvers are on our to-do list.
